java/vector/src/main/java/org/apache/arrow/vector/compression/AbstractCompressionCodec.java

```diff
@@ -0,0 +1,117 @@
+      compressedBuffer.writerIndex(CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH);
+      uncompressedBuffer.close();
+      return compressedBuffer;
+    }
+
+    ArrowBuf compressedBuffer = doCompress(allocator, uncompressedBuffer);
+    long compressedLength = compressedBuffer.writerIndex() - CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH;
+    long uncompressedLength = uncompressedBuffer.writerIndex();
+
+    if (compressedLength > uncompressedLength) {
+      // compressed buffer is larger, send the raw buffer
+      compressedBuffer.close();
+      compressedBuffer = CompressionUtil.packageRawBuffer(allocator, uncompressedBuffer);
+    } else {
+      writeUncompressedLength(compressedBuffer, uncompressedLength);
+    }
+
+    uncompressedBuffer.close();
+    return compressedBuffer;
+  }
+
+  @Override
+  public ArrowBuf decompress(BufferAllocator allocator, ArrowBuf compressedBuffer) {
+    Preconditions.checkArgument(compressedBuffer.writerIndex() <= Integer.MAX_VALUE,
+        "The compressed buffer size exceeds the integer limit");
+
+    Preconditions.checkArgument(compressedBuffer.writerIndex() >= CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH,
+        "Not enough data to decompress.");
+
+    long decompressedLength = compressedBuffer.getLong(0);
+    if (!MemoryUtil.LITTLE_ENDIAN) {
+      decompressedLength = Long.reverseBytes(decompressedLength);
+    }
+
+    if (decompressedLength == 0L) {
+      // shortcut for empty buffer
+      compressedBuffer.close();
+      return allocator.getEmpty();
+    }
+
+    if (decompressedLength == CompressionUtil.NO_COMPRESSION_LENGTH) {
+      // no compression
+      return CompressionUtil.extractUncompressedBuffer(compressedBuffer);
+    }
+
+    ArrowBuf decompressedBuffer = doDecompress(allocator, compressedBuffer);
+    compressedBuffer.close();
+    return decompressedBuffer;
+  }
+
+  protected void writeUncompressedLength(ArrowBuf compressedBuffer, long uncompressedLength) {
+    if (!MemoryUtil.LITTLE_ENDIAN) {
+      uncompressedLength = Long.reverseBytes(uncompressedLength);
+    }
+    // first 8 bytes reserved for uncompressed length, according to the specification
+    compressedBuffer.setLong(0, uncompressedLength);
+  }
+
+  protected long readUncompressedLength(ArrowBuf compressedBuffer) {
+    long decompressedLength = compressedBuffer.getLong(0);
+    if (!MemoryUtil.LITTLE_ENDIAN) {
+      decompressedLength = Long.reverseBytes(decompressedLength);
+    }
+    return decompressedLength;
+  }
+
+  /**
+   * The method that actually performs the data compression.
+   * Please note that the first 8 bytes of the compressed buffer is reserved for the uncompressed size.
+   */
+  protected abstract ArrowBuf doCompress(BufferAllocator allocator, ArrowBuf uncompressedBuffer);
+
+  /**
+   * The method that actually performs the data decompression.
+   * Please note that the first 8 bytes of the compressed buffer is reserved for the uncompressed size.
```

Comment:
I think for decompression it probably makes sense to provide the slice buffer only including the compressed data, and a preallocated output buffer to simply implementations of other codec.  What do you think?

Reply:
Either way lets clarify ownership of the passed in buffer.

Reply:
I have revised the JavaDoc to clarify the buffer ownership. Thanks.

The method of passing in only the compressed data sounds good to me. 
However, there are some problem when implementing the `doDecompress` method: it may also need the uncompressed size (e.g. to allocate some internal data structures, like on-heap array in our implementation). Such information can be obtained from the size of the passed-in buffer. However, the buffer size can be larger than the uncompressed size, which may cause waste of memory.

We can also change the interface to pass in the uncompressed size, but this would make the interface more complicated and asymmetric. 

So I have keeped the interface unchanged. If you feel strong about it, we can change the interface in that way.

Reply:
> Such information can be obtained from the size of the passed-in buffer. However, the buffer size can be larger than the uncompressed size, which may cause waste of memory.

In what cases would we pass in a buffer that is larger than the decompressed size? (it seems like it would could always be allocated to the exact size needed?  Or are you saying some buffers need to be larger then the simply the uncompressed size?)

Reply:
The default allocation policy is to round the requested size to the next power of 2. 
For example, if we ask for a 10MB buffer, the actual buffer size should be 16MB.

Reply:
but isn't that capacity and not actual size?  I don't think there is much of choice for consumers anyways since we are providing them with the BufferAllocator which has its policy fixed?

Reply:
Right. It is capacity. 
So we can use the `writerIndex` to denote the `actual size`. Does it sound good?

Reply:
I would like to try to avoid using writerIndex if possible.


What I was thinking originally, is that a new ArrowBuf with capacity equal to the decompressed size is passed in.  So implementations can simply reference buffer size to know uncompressed size, and can allocate additional buffers if needed for scratch. writerIndex would be zero (and ideally would not be touched by the implementations).  

At least looking at the zstd and lz4 c apis, it looks like both support this type of contract (extra output buffer padding doesn't seem necessary).

Reply:
Agreed with your points about `writerIndex`.
Do you have any suggestions concerning this API?

Reply:
I think so, would you mind if I push a commit to this PR? I'll try to make the changes this evening.

Reply:
Of course not. Please go ahead

Reply:
OK, I played around with this last night and couldn't get something to my satisfaction.  I don't think we should hold up merging this.